app/engine/utils/FullTSQuadraticSeries.js

```diff
@@ -0,0 +1,332 @@
+'use strict'
+/*
+  Open Rowing Monitor, https://github.com/laberning/openrowingmonitor
+
+  The TSLinearSeries is a datatype that represents a Linear Series. It allows
+  values to be retrieved (like a FiFo buffer, or Queue) but it also includes
+  a Theil-Sen Quadratic Regressor to determine the coefficients of this timeseries.
+
+  At creation its length is determined. After it is filled, the oldest will be pushed
+  out of the queue) automatically.
+
+  A key constraint is to prevent heavy calculations at the end (due to large
+  array based curve fitting), which might be performed on a Pi zero
+
+  The Theil-Senn implementation uses concepts that are described here:
+  https://stats.stackexchange.com/questions/317777/theil-sen-estimator-for-polynomial,
+
+  The determination of the coefficients is based on the math descirbed here:
+  https://www.quora.com/How-do-I-find-a-quadratic-equation-from-points/answer/Robert-Paxson,
+  https://www.physicsforums.com/threads/quadratic-equation-from-3-points.404174/
+
+  The array is ordered such that x[0] is the oldest, and x[x.length-1] is the youngest
+*/
+
+import { createSeries } from './Series.js'
+
+import loglevel from 'loglevel'
+const log = loglevel.getLogger('RowingEngine')
+
+function createTSQuadraticSeries (maxSeriesLength = 0) {
+  const X = createSeries(maxSeriesLength)
+  const Y = createSeries(maxSeriesLength)
+  const A = []
+  const B = []
+  const C = []
+  let _A = 0
+  let _B = 0
+  let _C = 0
+
+  function push (x, y) {
+    X.push(x)
+    Y.push(y)
+
+    if (maxSeriesLength > 0 && A.length >= maxSeriesLength) {
+      // The maximum of the array has been reached, we have to create room
+      // in the 2D array by removing the first row from the table
+      removeFirstRow()
+    }
+
+    // Invariant: the indices of the X and Y array now match up with the
+    // row numbers of the A array. So, the A of (X[0],Y[0]) and (X[1],Y[1]
+    // will be stored in A[0][.].
+
+    // Add an empty array at the end to store futurs results for the most recent points
+    A.push([])
+    B.push([])
+    C.push([])
+
+    // Calculate the coefficients of this new point
+    if (X.length() > 2) {
```

Comment:
Actually, the approach you propose would hurt the performance in an extremely bad way.

The calculation of a single coefficient (thus A, B, C) is something between O(N^2) and O(N^3), where N is the flanklength. This contains three major elements:
- The addition of the X and Y values to the array, which is O(1) operation.
- The second step is the generation of all credible A, B and C values based on all datapoints in the flank (which is O(N^2) to O(N^3), as you need to examine all combinations of datapoints in the flank). By using specific matrix-properties of Theil-Senn in a datastream, we can skip already known combinations (hence the matrices used for A, B and C), but for a new datapoint, we still need to calculate all credible A, B and C values for that new datapoint in combination with all previous datapoints, making it O(N^2).
- The third step is cryptic, but extremely CPU-expensive: when determining the matrix median, I need to serialize the matrix, sort it and determine the median. As the sort contains new values in all places in the matrix (due to the serialize) it is O(N^2). The out-commented code is more efficient and theoretically better, but it leads to unexplained behaviour when used on real-life data.

Combined, it will result in about O(N^3) calculations per new datapoint. This means that the calculation of a single coefficient with a flanklength of 10 will result in aprroximatly 1000 calculations. Depending on the call behaviour you can perform step 2 and 3 in a eager way (i.e. at datapoint addition) or in a lazy way (i.e. when the function is called).

Looking at the call behaviour, the Theil-Senn regressor is used quite intensely by `flywheel.js`, as angular velocity (first derived function, so y' = 2x * A + B), angular acceleration (second derived function, thus y'' = 2 A) need to calculated N times per datapoint. So calculation of these values by `flywheel.js` indirectly calls the calculation of A twice, and B once. As stroke detection and linear metrics are highly dependent on this, and multiple approximations of the same value need to be harmonized, these coefficients are calculated 3*N per datapoint.

Then you have three approaches:
- When the **current eager approach** is used, adding the datapoint from Flyhweel.js, will take O(N^3) CPU cycles, but all following calls are all O(1), making the total calculation used at Flywheel.js O(N) at retrieval. Here, performance is dominated by the O(N^3).
- A **partial lazy approach** where step 3 would be executed on demand, would still result in a O(N^2) to O(N^3) cost of addition of datapoint by `flywheel.js`, but it would require a O(N^2) calculation for each coefficient when called. Triggering this calculation 3 * N per new datapoint, making the retrieval part of `Flyhweel.js` 3 * O(N^3). Here, performance is dominated by the 3* O(N^3) of the retrieval, which is significantly worse than the eager approach.
- A **completely lazy approach**. performing step 2 and 3 on demand, would result in a O(1) cost of adding a datapoint. But retrieval is O(N^3), making a retrieval 3 * O(N^4). This is actually the worst performer of them all.

As you can see, any lazy approach would kill the performance of this regressor when combined with `Flywheel.js`. Please note, I run this on a Raspberry Pi 4 with flanklength 12, and it completes within the 5ms (minimum currentDt) and when running on simulated data a 1 hour row completes within 2 minutes, clearly indicating that CPU limits are no real hefty issues there. "Benefit" of the O(N^3) is that when used on a less data intensive machine, the CPU will shrink quickly: when using 5 datapoints, you end up with 125 calculations.

There is a simple version of Quadratic Regression, which isn't robust to outliers, which is O(1). But I haven't tested it as thoroughly as the current TS Quadratic regressor as the robustness of the algorithm is far less. I could provide it with an identical interface and provide it, but I also feel that more options for algorithms tends to add a lot of complexity for both users and maintainers, so I rather only add it when there is a clear need for it.